File: stack/java/NestingScore.java
import java.util.ArrayDeque;
import java.util.Deque;

public final class NestingScore {
  private NestingScore() {
  }

  public static int score(String brackets) {
    Deque<Integer> stack = new ArrayDeque<>();
    stack.push(0);
    for (char ch : brackets.toCharArray()) {
      if (ch == '[') {
        stack.push(0);
      } else {
        int top = stack.pop();
        int value = top == 0 ? 1 : 2 * top;
        int updated = stack.pop() + value;
        stack.push(updated);
      }
    }
    return stack.peek();
  }
}

File: stack/ts/nestingScore.ts
export function nestingScore(brackets: string): number {
  const stack: number[] = [0];
  for (const ch of brackets) {
    if (ch === '[') {
      stack.push(0);
    } else {
      const top = stack.pop() ?? 0;
      if (top === 0) {
        stack[stack.length - 1] += 1;
      } else {
        stack[stack.length - 1] += 2 * top;
      }
    }
  }
  return stack[0];
}
